libdrgn/arch_x86_64.c

```diff
@@ -243,6 +243,54 @@ get_registers_from_frame_pointer(struct drgn_program *prog,
 	return NULL;
 }
 
+/**
+ * Escapes the effect a call operation has on registers, changing registers
+ * so that RIP points to the caller frame post-call instruction and RSP is
+ * increased to point to the top of the caller frame.
+ * This can be used in cases where, for example, a NULL function pointer
+ * was called and as a result the RIP (program counter) value of zero prevents
+ * any unwinding.
+ */
+static struct drgn_error *
+unwind_call(struct drgn_program *prog,
+		       struct drgn_register_state *regs,
+		       struct drgn_register_state **ret)
+{
+	struct drgn_error *err;
+
+	struct optional_uint64 rsp =
+		drgn_register_state_get_u64(prog, regs, rsp);
+	if (!rsp.has_value)
+		return &drgn_stop;
+
+	/*
+	 * Read RSP which contains the return address to be used
+	 * as the new RIP value.
+	 */
+	uint64_t postcall_rip;
+	err = drgn_program_read_memory(prog, &postcall_rip, rsp.value,
+		sizeof(postcall_rip), false);
+	if (err) {
+		if (err->code == DRGN_ERROR_FAULT) {
+			drgn_error_destroy(err);
+			err = &drgn_stop;
+		}
+		return err;
+	}
+
+	struct drgn_register_state *tmp = drgn_register_state_dup(regs);
```

Comment:
Need to check that the allocation succeeded.

```suggestion
	struct drgn_register_state *tmp = drgn_register_state_dup(regs);
	if (!tmp)
		return &drgn_enomem;
```